tools/traffic/workers/blob_reader.go

```diff
@@ -0,0 +1,223 @@
+		recombinationFailureMetric: generatorMetrics.NewCountMetric("recombination_failure"),
+		readLatencyMetric:          generatorMetrics.NewLatencyMetric("read"),
+		validBlobMetric:            generatorMetrics.NewCountMetric("valid_blob"),
+		invalidBlobMetric:          generatorMetrics.NewCountMetric("invalid_blob"),
+		readSuccessMetric:          generatorMetrics.NewCountMetric("read_success"),
+		readFailureMetric:          generatorMetrics.NewCountMetric("read_failure"),
+		operatorSuccessMetrics:     make(map[core.OperatorID]metrics.CountMetric),
+		operatorFailureMetrics:     make(map[core.OperatorID]metrics.CountMetric),
+		requiredReadPoolSizeMetric: generatorMetrics.NewGaugeMetric("required_read_pool_size"),
+		optionalReadPoolSizeMetric: generatorMetrics.NewGaugeMetric("optional_read_pool_size"),
+	}
+}
+
+// Start begins a blob reader goroutine.
+func (reader *BlobReader) Start() {
+	reader.waitGroup.Add(1)
+	ticker := time.NewTicker(reader.config.ReadRequestInterval)
+	go func() {
+		defer reader.waitGroup.Done()
+		for {
+			select {
+			case <-(*reader.ctx).Done():
+				return
+			case <-ticker.C:
+				reader.randomRead()
+			}
+		}
+	}()
+}
+
+// randomRead reads a random blob.
+func (reader *BlobReader) randomRead() {
+	metadata := reader.blobsToRead.GetNext()
+	if metadata == nil {
+		// There are no blobs that we are required to read.
+		return
+	}
+
+	reader.requiredReadPoolSizeMetric.Set(float64(reader.blobsToRead.Size()))
+
+	ctxTimeout, cancel := context.WithTimeout(*reader.ctx, reader.config.FetchBatchHeaderTimeout)
+	batchHeader, err := metrics.InvokeAndReportLatency(reader.fetchBatchHeaderMetric,
+		func() (*contractEigenDAServiceManager.IEigenDAServiceManagerBatchHeader, error) {
+			return reader.chainClient.FetchBatchHeader(
+				ctxTimeout,
+				gcommon.HexToAddress(reader.config.EigenDAServiceManager),
+				metadata.BatchHeaderHash[:],
+				big.NewInt(int64(0)),
+				nil)
+		})
+	cancel()
+	if err != nil {
+		reader.logger.Error("failed to get batch header", "err:", err)
+		reader.fetchBatchHeaderFailure.Increment()
+		return
+	}
+	reader.fetchBatchHeaderSuccess.Increment()
+
+	var batchHeaderHash [32]byte
+	copy(batchHeaderHash[:], metadata.BatchHeaderHash[:])
+
+	ctxTimeout, cancel = context.WithTimeout(*reader.ctx, reader.config.RetrieveBlobChunksTimeout)
+	chunks, err := metrics.InvokeAndReportLatency(reader.readLatencyMetric, func() (*clients.BlobChunks, error) {
+		return reader.retriever.RetrieveBlobChunks(
+			ctxTimeout,
+			batchHeaderHash,
+			uint32(metadata.BlobIndex),
+			uint(batchHeader.ReferenceBlockNumber),
+			batchHeader.BlobHeadersRoot,
+			core.QuorumID(0))
+	})
+	cancel()
+	if err != nil {
+		reader.logger.Error("failed to read chunks", "err:", err)
+		reader.readFailureMetric.Increment()
+		return
+	}
+	reader.readSuccessMetric.Increment()
+
+	assignments := chunks.Assignments
+
+	data, err := reader.retriever.CombineChunks(chunks)
+	if err != nil {
```

Comment:
This is an interesting case and may warrant different error handling vs. the read failure above.
Read failures may happen for different reasons but mostly suggest some issues with network health. 
Blob recovery failure here most certainly suggests a bug. So we may set up noisier alert for this one

Reply:
This makes an error log and updates a metric. What did you have in mind for a noisier alert?